File: scripts/scrape_un2720.py
#!/usr/bin/env python3
"""
Scrape UN2720 collected data JSON API and output a normalized JSON file.
Run: pip install -r requirements-scrape.txt && python scripts/scrape_un2720.py
Env: GCS_BUCKET, GCS_OBJECT (e.g. un2720.json) to upload; otherwise prints to stdout.
"""
from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from typing import Any, Dict, List

import requests

API_URL = "https://app.un2720.org/fetchCollectedData"


def parse_num(value: Any) -> int | float:
    if value is None:
        return 0
    s = str(value)
    s = s.replace(" ", "").replace(",", "")
    try:
        # Some fields are floats (e.g. weightTonnes)
        return float(s)
    except (ValueError, TypeError):
        return 0


def scrape() -> Dict[str, Any]:
    # Match dashboard date range behaviour (full history until today)
    params = {
        "start": "2025-05-19",
        "end": datetime.now(timezone.utc).strftime("%Y-%m-%d"),
    }
    headers = {
        "referer": "https://app.un2720.org/tracking/collected",
        "user-agent": "ESPA-Israel-Scraper/1.0 (+https://www.espa-israel.com)",
        "accept": "application/json, text/plain, */*",
    }

    resp = requests.get(API_URL, params=params, headers=headers, timeout=60)
    resp.raise_for_status()
    raw = resp.json()

    # Top-level summary cards (exact values from API)
    summary = {
        "pallets": int(parse_num(raw.get("totalPallets"))),
        "trucks": int(parse_num(raw.get("totalTrucks"))),
        "weightTonnes": float(parse_num(raw.get("totalWeight"))),
        "numberOfRequests": int(parse_num(raw.get("totalUtns"))),
    }

    # Organizations: prefer orgTableData (normalized objects). Fallback to label/count arrays.
    orgs: List[Dict[str, Any]] = []

    table_data = raw.get("orgTableData")
    if isinstance(table_data, list) and table_data:
        for row in table_data:
            if not isinstance(row, dict):
                continue
            name = row.get("organization") or row.get("name") or row.get("org")
            if not name:
                continue
            orgs.append(
                {
                    "organization": str(name),
                    "trucks": int(parse_num(row.get("trucks") or row.get("truckCount"))),
                    "pallets": int(parse_num(row.get("pallets") or row.get("palletCount"))),
                    "weightTonnes": float(parse_num(row.get("weight") or row.get("weightTonnes"))),
                    "numberOfRequests": int(parse_num(row.get("utns") or row.get("requests") or 0)),
                }
            )
    else:
        labels = raw.get("orgLabels") or []
        truck_counts = raw.get("orgTruckCounts") or []
        pallet_counts = raw.get("orgPalletCounts") or []
        for idx, name in enumerate(labels):
            trucks = truck_counts[idx] if idx < len(truck_counts) else 0
            pallets = pallet_counts[idx] if idx < len(pallet_counts) else 0
            orgs.append(
                {
                    "organization": str(name),
                    "trucks": int(parse_num(trucks)),
                    "pallets": int(parse_num(pallets)),
                    # Weight / requests not directly available per org in this shape
                    "weightTonnes": 0.0,
                    "numberOfRequests": 0,
                }
            )

    # Attach full API payload for charts (daily trends, commodity pie, crossings, etc.)
    api_payloads = [
        {
            "url": API_URL,
            "body": raw,
        }
    ]

    return {
        "summary": summary,
        "organizations": orgs,
        "fetchedAt": datetime.now(tz=timezone.utc).isoformat(),
        "apiPayloads": api_payloads,
    }


def main() -> None:
    data = scrape()
    payload = json.dumps(data, indent=2)

    bucket = os.environ.get("GCS_BUCKET")
    obj = os.environ.get("GCS_OBJECT", "un2720.json")
    if bucket:
        try:
            from google.cloud import storage

            client = storage.Client()
            blob = client.bucket(bucket).blob(obj)
            blob.upload_from_string(payload, content_type="application/json")
            # Public read: use bucket IAM (allUsers:objectViewer) if allowed by org policy
            print(f"Uploaded gs://{bucket}/{obj}", file=sys.stderr)
        except Exception as e:
            print(f"GCS upload failed: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        print(payload)


if __name__ == "__main__":
    main()

File: scripts/scrape-un2720.mjs
#!/usr/bin/env node
/**
 * Scrapes UN2720 tracking/collected dashboard (client-rendered) and outputs JSON.
 * Run: node scripts/scrape-un2720.mjs [output-path]
 * Output path default: public/data/un2720.json
 * Requires: npm install puppeteer (or run via GitHub Action with puppeteer in deps).
 */

import { launch } from "puppeteer";
import { writeFileSync, mkdirSync } from "fs";
import { dirname } from "path";

const URL = "https://app.un2720.org/tracking/collected";
const DEFAULT_OUT = "public/data/un2720.json";

async function scrape() {
  const browser = await launch({
    headless: true,
    args: ["--no-sandbox", "--disable-setuid-sandbox"],
  });
  try {
    const page = await browser.newPage();
    const apiPayloads = [];

    // Capture JSON API responses that back the dashboard charts (crossings, commodities, daily trends, etc.)
    page.on("response", async (response) => {
      try {
        const url = response.url();
        if (!url.startsWith("https://app.un2720.org/")) return;
        const contentType = (response.headers()["content-type"] || "").toLowerCase();
        if (!contentType.includes("application/json")) return;
        const body = await response.json().catch(() => null);
        if (!body) return;
        apiPayloads.push({ url, body });
      } catch {
        // Ignore capture errors; scraping must remain resilient
      }
    });
    await page.setUserAgent("ESPA-Israel-Scraper/1.0 (+https://www.espa-israel.com)");
    await page.goto(URL, { waitUntil: "networkidle2", timeout: 60000 });

    // Wait for dashboard content (stats or table)
    await page.waitForSelector("table, [class*='table'], h2, h3, [class*='summary']", {
      visible: true,
      timeout: 15000,
    }).catch(() => null);
    await new Promise((r) => setTimeout(r, 3000));

    const data = await page.evaluate(() => {
      const orgRows = [];

      const parseNum = (s) => {
        const n = parseInt(String(s).replace(/\s|,/g, ""), 10);
        return Number.isNaN(n) ? 0 : n;
      };

      // Table: find table and extract rows (Organization, Trucks, Pallets, Weight (t), Number of Requests)
      const table = document.querySelector("table");
      if (table) {
        const rows = Array.from(table.querySelectorAll("tbody tr"));
        const headerCells = table.querySelectorAll("thead th");
        const headerTexts = Array.from(headerCells).map((th) => th.innerText.trim().toLowerCase());
        const colOrg = headerTexts.findIndex((h) => h.includes("organization") || h.includes("org"));
        const colTrucks = headerTexts.findIndex((h) => h.includes("truck"));
        const colPallets = headerTexts.findIndex((h) => h.includes("pallet"));
        const colWeight = headerTexts.findIndex((h) => h.includes("weight"));
        const colRequests = headerTexts.findIndex((h) => h.includes("request"));

        for (const tr of rows) {
          const cells = tr.querySelectorAll("td");
          if (cells.length < 2) continue;
          const get = (idx) => (idx >= 0 && cells[idx] ? cells[idx].innerText.trim() : "");
          orgRows.push({
            organization: get(colOrg >= 0 ? colOrg : 0),
            trucks: parseNum(get(colTrucks >= 0 ? colTrucks : 1)),
            pallets: parseNum(get(colPallets >= 0 ? colPallets : 2)),
            weightTonnes: parseNum(get(colWeight >= 0 ? colWeight : 3)),
            numberOfRequests: parseNum(get(colRequests >= 0 ? colRequests : 4)),
          });
        }
      }

      const summary = {
        pallets: orgRows.reduce((sum, r) => sum + r.pallets, 0),
        trucks: orgRows.reduce((sum, r) => sum + r.trucks, 0),
        weightTonnes: orgRows.reduce((sum, r) => sum + r.weightTonnes, 0),
        numberOfRequests: orgRows.reduce((sum, r) => sum + r.numberOfRequests, 0),
      };

      return { summary, organizations: orgRows };
    });

    const out = {
      summary: data.summary,
      organizations: data.organizations.filter((r) => r.organization || r.trucks || r.pallets),
      fetchedAt: new Date().toISOString(),
      // Raw API payloads include data for:
      // - weight by crossing
      // - collected weight by commodity
      // - collected daily trends (pallets / trucks / weight over time)
      apiPayloads,
    };
    return out;
  } finally {
    await browser.close();
  }
}

const outPath = process.argv[2] || DEFAULT_OUT;
scrape()
  .then((json) => {
    mkdirSync(dirname(outPath), { recursive: true });
    writeFileSync(outPath, JSON.stringify(json, null, 2), "utf8");
    console.log("Wrote", outPath);
  })
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
